common/events/platform/boost/events.cpp

```diff
@@ -51,6 +56,55 @@ void Timer::Cancel() {
 	timer_.cancel();
 }
 
+void SignalHandler::Cancel() {
+	signal_set_.cancel();
+}
+
+error::Error SignalHandler::AsyncWait(const SignalSet &set, SignalHandlerFn handler_fn) {
+	signal_set_.clear();
+	for (auto sig_num : set) {
+		boost::system::error_code ec;
+		signal_set_.add(sig_num, ec);
+		if (ec) {
+			return error::Error(
+				ec.default_error_condition(),
+				"Could not add signal " + std::to_string(sig_num) + " to signal set");
+		}
+	}
+
+	class SignalHandlerFunctor {
+	public:
+		asio::signal_set &sig_set;
+		SignalHandlerFn handler_fn;
+
+		void operator()(const boost::system::error_code &ec, int signal_number) {
+			if (ec) {
+				if (ec == boost::asio::error::operation_aborted) {
+					// All handlers are called with this error when the signal set
+					// is cancelled.
+					return;
+				}
+				log::Error("Failure in signal handler: " + ec.message());
+			} else {
+				handler_fn(signal_number);
+			}
+
+			// in either case, register ourselves again because
+			// asio::signal_set::async_wait() is a one-off thing
+			sig_set.async_wait(*this);
+		}
+	};
+
+	SignalHandlerFunctor fctor {signal_set_, handler_fn};
+	signal_set_.async_wait(fctor);
+
+	// Unfortunately, asio::signal_set::async_wait() doesn't return anything so
+	// we have no chance to propagate any error here (wonder what they do if
+	// they get an error from sigaction() or whatever they use
+	// internally). Other implementations may have better options, though.
```

Comment:
Omg, you're right, we're even using it! I was confusing it with our handler, which doesn't. So maybe that's the answer: Any errors in `sigaction` are delivered through the error code in the handler.

Reply:
Depends on when it does `sigaction`. Do you think it only happens at `loop.Run()`? Because otherwise the handler is not called until that happens. It's still quite weird to me.